### PR TITLE
Retry answers functionality

### PR DESCRIPTION
**Pull request by jmoreira-valory**
Retry answers functionality

### CODE REVIEW

File: packages/valory/services/market_maker/service.yaml [hunk 1]
```diff
@@ -71,6 +71,8 @@ models:
       multisend_batch_size: ${MULTISEND_BATCH_SIZE:int:1}
       ipfs_address: ${IPFS_ADDRESS:str:https://gateway.autonolas.tech/ipfs/}
       mech_contract_address: ${MECH_CONTRACT_ADDRESS:str:0x77af31de935740567cf4ff1986d04b2c964a786a}
+      answer_retry_intervals: ${ANSWER_RETRY_INTERVALS:list:[0, 86400, 259200, 604800,
+        1209600]}
```
Comment (Adamantios) [inline packages/valory/services/market_maker/service.yaml:75]: Wouldn't it be simpler if you implemented a backoff logic and made the backoff factor configurable?

File: packages/valory/skills/market_creation_manager_abci/behaviours.py [hunk 2]
```diff
@@ -2291,6 +2291,54 @@ def _get_balance(self, account: str) -> Generator[None, None, Optional[int]]:
...
+
+            if question_id in self.shared_state.questions_responded:
+                continue
+
+            if question_id not in self.shared_state.questions_requested_mech:
+                self.shared_state.questions_requested_mech[question_id] = {}
+                self.shared_state.questions_requested_mech[question_id][
+                    "question"
+                ] = question
+                self.shared_state.questions_requested_mech[question_id]["retries"] = []
```
Comment (Adamantios) [inline packages/valory/skills/market_creation_manager_abci/behaviours.py:2316]:
```suggestion
                self.shared_state.questions_requested_mech[question_id] = {
                    "question": question,
                    "retries": []
                }
```

File: packages/valory/skills/market_creation_manager_abci/behaviours.py [hunk 3]
```diff
@@ -2291,6 +2291,54 @@ def _get_balance(self, account: str) -> Generator[None, None, Optional[int]]:
...
+
+            retries = self.shared_state.questions_requested_mech[question_id]["retries"]
+            n_retries = len(retries)
+            time_since_last_retry = now - retries[-1] if retries else 0
+            retry_period = self.params.answer_retry_intervals[
+                min(n_retries, len(self.params.answer_retry_intervals) - 1)
+            ]
```
Comment (Adamantios) [inline packages/valory/skills/market_creation_manager_abci/behaviours.py:2323]: Hard to read. Why not pop from a local copy instead?

File: packages/valory/skills/market_creation_manager_abci/behaviours.py [hunk 4]
```diff
@@ -2291,6 +2291,54 @@ def _get_balance(self, account: str) -> Generator[None, None, Optional[int]]:
...
+
+        self.context.logger.info(
+            f"Determined {len(eligible_questions_id)} eligible questions to answer."
+        )
+
+        num_questions = min(
+            len(eligible_questions_id), self.params.multisend_batch_size
+        )
+        random.seed(self.last_synced_timestamp)
+        random_questions_id = random.sample(eligible_questions_id, num_questions)
```
Comment (Adamantios) [inline packages/valory/skills/market_creation_manager_abci/behaviours.py:2335]: Randomness should always come from an external source.

Comment (jmoreira-valory) [reply packages/valory/skills/market_creation_manager_abci/behaviours.py:2335]: It comes from the synchronized timestamp of all the agents.

**Review by Adamantios — APPROVED** [review]

File: packages/valory/skills/market_creation_manager_abci/behaviours.py (hunk 3 above)
Comment (jmoreira-valory) [reply packages/valory/skills/market_creation_manager_abci/behaviours.py:2322]: Fixed

File: packages/valory/skills/market_creation_manager_abci/behaviours.py (hunk 4 above)
Comment (jmoreira-valory) [reply packages/valory/skills/market_creation_manager_abci/behaviours.py:2333]: The collect randomness is after this state, so this need refactor. Opened an issue for this: https://github.com/valory-xyz/market-creator/issues/103

**Pull request merged**